thunder/executors/cudnn_sdpa.py

```diff
@@ -0,0 +1,759 @@
+    value = _sdpa_enforce_input_tensor_contiguity(value)
+
+    # When cat_grad_qkv is on, allocate dQKV and make dQ, dK, and dV
+    # slices of that. Otherwise, allocate them individually.
+    grad_qkv: None | torch.Tensor = None
+    if cat_grad_qkv:
+        grad_qkv = _allocate_catted_grad_qkv(query, key, value)
+        grad_query, grad_key, grad_value = grad_qkv.split([query.size(1), key.size(1), value.size(1)], dim=1)
+    else:
+        grad_query = torch.empty_like(query)
+        grad_key = torch.empty_like(key)
+        grad_value = torch.empty_like(value)
+
+    if attn_mask is not None:
+        if query.ndim > attn_mask.ndim:
+            attn_mask = attn_mask.view(*((1,) * (query.ndim - attn_mask.ndim)), *attn_mask.shape)
+        if attn_mask.dtype == torch.bool:
+            attn_bias = torch.zeros_like(attn_mask, dtype=query.dtype)
+            attn_bias.masked_fill_(attn_mask.logical_not(), float("-inf"))
+            attn_mask = attn_bias
+
+    (
+        Q,
+        K,
+        V,
+        O,
+        dO,
+        Stats,
+        Seed,
+        Offset,
+        Attn_scale,
+        Bias,
+        dQ,
+        dK,
+        dV,
+        dBias,
+        graph,
+    ) = _make_cudnn_sdpa_backward_graph(
+        query,
+        key,
+        value,
+        attn_mask,
+        dropout_p,
+        is_causal,
+        query.stride(),
+        key.stride(),
+        value.stride(),
+        grad_query.stride(),
+        grad_key.stride(),
+        grad_value.stride(),
+    )
+
+    # Default value of scale, if not provided, in all torch versions
+    if scale is None:
+        scale = query.shape[-1] ** -0.5
+    Attn_scale_cpu = torch.full((1, 1, 1, 1), scale, dtype=torch.float32, device="cpu")
+
+    cudnn_to_torch_tensor = {
+        dO: grad_out.detach(),
+        Q: query.detach(),
+        K: key.detach(),
+        V: value.detach(),
+        Attn_scale: Attn_scale_cpu,
+        O: out.detach(),
+        Stats: softmax_stats,
+        Seed: philox_seed,
+        Offset: philox_offset,
+        dQ: grad_query,
+        dK: grad_key,
+        dV: grad_value,
+    }
+    if attn_mask is not None:
+        grad_attn_mask = torch.empty_like(attn_mask)
+
+        cudnn_to_torch_tensor[Bias] = attn_mask.detach()
+        cudnn_to_torch_tensor[dBias] = grad_attn_mask
+
+    workspace = torch.empty(graph.get_workspace_size(), device=query.device, dtype=torch.uint8)
+
+    # Even though the handle is created on query.device, cudnn still requires to set current device to query.device.
+    # This is most probably a bug and is being actively looked into.
+    with torch.cuda.device(query.device):
+        graph.execute(cudnn_to_torch_tensor, workspace, handle=_get_cudnn_handle(query.device))
+
+    if cat_grad_qkv:
+        grads = (grad_qkv,)
+    else:
+        grads = (grad_query, grad_key, grad_value)
+
+    if attn_mask is not None:
+        grads = grads + (grad_attn_mask,)
+    return grads
+
+
```

Comment:
Let's keep the registration here because the symbol is needed later on


```suggestion
cudnn_sdpa_bwd = cudnn_ex.register_operator(
    "cudnn_sdpa_bwd",
    meta=_cudnn_sdpa_bwd_meta,
    fn=_cudnn_sdpa_bwd_impl,
)


```